Exploded entities issues fix (#819)

xanderdin opened the pull request:
I believe this fixes that old issue. Tested with different combinations of blocks, including different layers, different colors, line types, widths, blocks inside blocks to several insertion levels. Encountered no side effects or differences in exploded results.

xanderdin commented:
Oh, the things you described happened to be unexpected to me. I tested undo but not after file reopen. Should I close the pull request for now (or you just reject it?) until I get more time to look for a better solution?

lordofbikes commented:
Don't worry, this happens. I only found this by chance.
You can leave this open if you can fix this approach.
We can still close it later, if you find a better solution or prefer to start a new branch for this.

xanderdin commented:
Just to be clear. The issue with file reopening is not consequence of my fix. The same behavior is in the master, and I think it is rather another issue, not the one described in #819. Still don't know how to fix it.

lordofbikes commented:
As I said, I found this by accident on my code review, but, to be honest, didn't cross check with master.
Hopefully I find some time on the weekend to have a deeper look at your fix and this issue.
Many thanks for your efforts!

xanderdin commented:
Meanwhile, let me explain my fix and my decisions. Hope it will help during deeper code review.

As described in #819, when exploding a rectangle only one line out of four retains properties of its parent container. And that line is the first one. So it is pretty clear the others should get its properties the same way as the first does. That's the main idea.

I found out quickly that `resolvePen` variable in explode() is always false. In LibreCAD 2.1.3 exploded entities alway were getting their pen properties from their container by the line of code (notice **ec**):

https://github.com/LibreCAD/LibreCAD/blob/f21fde6e5ee4a63b68e72ad59a3fa4268dbe1314/librecad/src/lib/modification/rs_modification.cpp#L3018

After https://github.com/LibreCAD/LibreCAD/commit/e3c6e6ca73fae7434bb53940c4c26229239b43e1 fix, exploded entities are getting their pen properties always from themselves (notice **e2**), not their parent container (resolvePen is always false).

https://github.com/LibreCAD/LibreCAD/blob/8604f171ee380f294102da6154adf77ab754d403/librecad/src/lib/modification/rs_modification.cpp#L3093-L3097

By changing resolvePen to true the issue seems to be fixed... for a simple container, e.g. rectangle or a single polyline. But what if a container contains a block of rectangles? Then, when exploded those rectangles all were getting the same deficiency with only one line retained pen properties.

Thus I got the idea of traversal down the tree of exploded entities and updating their properties for all descendants of exploded block.

As far as code for setting properties is the same for all descendants, I decided to move it from explode() to a single function which calls itself every time another container is found among descendants.

Among other lines of code I moved from explode() to update_exploded_children_recursively() is the following line of code I changed.

Was:
```
clone->setPen(ec->getPen(true));
```
I changed to:
```
clone->setPen(ec->getPen(false));
```

The reason for the change is, the former line in some cases produces bad side effects, but the latter is not. Because of the former line of code was never called (resolvePen was always false) but the latter was the same as in 2.1.3 code before the https://github.com/LibreCAD/LibreCAD/commit/e3c6e6ca73fae7434bb53940c4c26229239b43e1 fix was applied, I decided to change it as shown above.

After those changes everything became almost ok, but in some situations some entities was displayed on the screen with wrong pen properties. Happily, I noticed when clicking on 'eye' icon near any layer, toggling the layer visibility, was changing that wrong display to the right one. Investigating that issue I found that toggling layer visibility calls container->updateInserts(). (By the way, container->updateInserts() is called only when toggling single layer visibility, and not called when 'Hide/Unhide all layers' button pressed. Is that ok or a bug?) So I added that call also to the end of explode(). After that it seems everything works as expected.

I feel like parameter names of the update_exploded_children_recursively() may look somewhat misleading. But I wanted to retain their meanings from the explode(). Anyway, any suggestions are welcomed.

lordofbikes commented:
Many thanks, also for creating #1049
